### PR TITLE
[Merged by Bors] - feat: port Algebra.Associated

### PR DESCRIPTION
**Pull request by siddhartha-gadgil**
dcf2250875895376a142faeeac5eabff32c48655

* [x] depends on #1055
* [x] depends on #1092

### CODE REVIEW

File: Mathlib/Algebra/Associated.lean [hunk 1]
```diff
@@ -0,0 +1,1262 @@
...
+    rw [pow_one]
+    rw [pow_one] at hxy
+    exact (h.dvd_or_dvd hxy).resolve_right hy
+  rw [pow_succ] at hxy⊢
+  obtain ⟨x', rfl⟩ := (h.dvd_or_dvd (dvd_of_mul_right_dvd hxy)).resolve_right hy
+  rw [mul_assoc] at hxy
+  exact mul_dvd_mul_left p (ih ((mul_dvd_mul_iff_left h.ne_zero).mp hxy))
+#align prime_pow_succ_dvd_mul prime_pow_succ_dvd_mul
+
+/-- `irreducible p` states that `p` is non-unit and only factors into units.
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:166]:
```suggestion
/-- `Irreducible p` states that `p` is non-unit and only factors into units.
```

File: Mathlib/Algebra/Associated.lean [hunk 2]
```diff
@@ -0,0 +1,1262 @@
...
+
+We explicitly avoid stating that `p` is non-zero, this would require a semiring. Assuming only a
+monoid allows us to reuse irreducible for associated elements.
+-/
+structure Irreducible [Monoid α] (p : α) : Prop where
+  /-- `p` is not a unit -/
+  not_unit : ¬IsUnit p
+  /-- if `p` factors then one factor is a unit -/
+  is_unit_or_is_unit' : ∀ a b, p = a * b → IsUnit a ∨ IsUnit b
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:175]:
```suggestion
  isUnit_or_isUnit' : ∀ a b, p = a * b → IsUnit a ∨ IsUnit b
```
also elsewhere

File: Mathlib/Algebra/Associated.lean [hunk 3]
```diff
@@ -0,0 +1,1262 @@
...
+    rw [mul_assoc, ← HAB]
+  · rw [← a⁻¹.isUnit_units_mul]
+    apply h
+    rw [mul_assoc, ← HAB, Units.inv_mul_cancel_left]
+#align irreducible_units_mul irreducible_units_mul
+
+theorem irreducible_is_unit_mul {a b : α} (h : IsUnit a) : Irreducible (a * b) ↔ Irreducible b :=
+  let ⟨a, ha⟩ := h
+  ha ▸ irreducible_units_mul a b
+#align irreducible_is_unit_mul irreducible_is_unit_mul
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:269]:
```suggestion
theorem irreducible_isUnit_mul {a b : α} (h : IsUnit a) : Irreducible (a * b) ↔ Irreducible b :=
  let ⟨a, ha⟩ := h
  ha ▸ irreducible_units_mul a b
#align irreducible_is_unit_mul irreducible_isUnit_mul
```

File: Mathlib/Algebra/Associated.lean [hunk 4]
```diff
@@ -0,0 +1,1262 @@
...
+    rw [← mul_assoc, ← HAB]
+  · rw [← Units.isUnit_mul_units B a⁻¹]
+    apply h
+    rw [← mul_assoc, ← HAB, Units.mul_inv_cancel_right]
+#align irreducible_mul_units irreducible_mul_units
+
+theorem irreducible_mul_is_unit {a b : α} (h : IsUnit a) : Irreducible (b * a) ↔ Irreducible b :=
+  let ⟨a, ha⟩ := h
+  ha ▸ irreducible_mul_units a b
+#align irreducible_mul_is_unit irreducible_mul_is_unit
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:285]:
```suggestion
theorem irreducible_mul_isUnit {a b : α} (h : IsUnit a) : Irreducible (b * a) ↔ Irreducible b :=
  let ⟨a, ha⟩ := h
  ha ▸ irreducible_mul_units a b
#align irreducible_mul_is_unit irreducible_mul_isUnit
```

File: Mathlib/Algebra/Associated.lean [hunk 5]
```diff
@@ -0,0 +1,1262 @@
...
+
+end CancelCommMonoidWithZero
+
+/-- Two elements of a `monoid` are `associated` if one of them is another one
+multiplied by a unit on the right. -/
+def Associated [Monoid α] (x y : α) : Prop :=
+  ∃ u : αˣ, x * u = y
+#align associated Associated
+
+-- mathport name: «expr ~ᵤ »
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:384]:
```suggestion
```

File: Mathlib/Algebra/Associated.lean [hunk 6]
```diff
@@ -0,0 +1,1262 @@
...
+
+theorem associated_unit_mul_right {β : Type _} [CommMonoid β] (a u : β) (hu : IsUnit u) :
+    Associated a (u * a) :=
+  (associated_unit_mul_left a u hu).symm
+#align associated_unit_mul_right associated_unit_mul_right
+
+theorem associated_mul_is_unit_left_iff {β : Type _} [Monoid β] {a u b : β} (hu : IsUnit u) :
+    Associated (a * u) b ↔ Associated a b :=
+  ⟨trans (associated_mul_unit_right _ _ hu), trans (associated_mul_unit_left _ _ hu)⟩
+#align associated_mul_is_unit_left_iff associated_mul_is_unit_left_iff
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:483]:
```suggestion
theorem associated_mul_isUnit_left_iff {β : Type _} [Monoid β] {a u b : β} (hu : IsUnit u) :
    Associated (a * u) b ↔ Associated a b :=
  ⟨trans (associated_mul_unit_right _ _ hu), trans (associated_mul_unit_left _ _ hu)⟩
#align associated_mul_is_unit_left_iff associated_mul_isUnit_left_iff
```
and a few more times below

File: Mathlib/Algebra/Associated.lean [hunk 7]
```diff
@@ -0,0 +1,1262 @@
...
+        Quotient.sound <| associated_one_of_associated_mul_one <| by rwa [mul_comm] at this⟩)
+    (by simp (config := { contextual := true }))
+#align associates.mul_eq_one_iff Associates.mul_eq_one_iff
+
+theorem units_eq_one (u : (Associates α)ˣ) : u = 1 :=
+  Units.ext (mul_eq_one_iff.1 u.val_inv).1
+#align associates.units_eq_one Associates.units_eq_one
+
+instance uniqueUnits : Unique (Associates
+          α)ˣ where
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:884]:
```suggestion
instance uniqueUnits : Unique (Associates α)ˣ where
```

File: Mathlib/Algebra/Associated.lean [hunk 8]
```diff
@@ -0,0 +1,1262 @@
...
+  default := 1
+  uniq := Associates.units_eq_one
+#align associates.unique_units Associates.uniqueUnits
+
+theorem coe_unit_eq_one (u : (Associates α)ˣ) : (u : Associates α) = 1 := by simp
+#align associates.coe_unit_eq_one Associates.coe_unit_eq_one
+
+theorem is_unit_iff_eq_one (a : Associates α) : IsUnit a ↔ a = 1 :=
+  Iff.intro (fun ⟨_, h⟩ => h ▸ coe_unit_eq_one _) fun h => h.symm ▸ isUnit_one
+#align associates.is_unit_iff_eq_one Associates.is_unit_iff_eq_one
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:894]:
```suggestion
theorem isUnit_iff_eq_one (a : Associates α) : IsUnit a ↔ a = 1 :=
  Iff.intro (fun ⟨_, h⟩ => h ▸ coe_unit_eq_one _) fun h => h.symm ▸ isUnit_one
#align associates.is_unit_iff_eq_one Associates.isUnit_iff_eq_one
```

File: Mathlib/Algebra/Associated.lean [hunk 9]
```diff
@@ -0,0 +1,1262 @@
...
+#align associates.exists_non_zero_rep Associates.exists_non_zero_rep
+
+end MonoidWithZero
+
+section CommMonoidWithZero
+
+variable [CommMonoidWithZero α]
+
+instance : CommMonoidWithZero (Associates α) where
+    -- Associates.CommMonoid, Associates.hasZero with
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:998]:
```suggestion
```

File: Mathlib/Algebra/Associated.lean [hunk 10]
```diff
@@ -0,0 +1,1262 @@
...
+      rintro ⟨a⟩
+      show Associates.mk (a * 0) = Associates.mk 0
+      rw [mul_zero]
+
+instance : OrderTop (Associates α) where
+  top := 0
+  le_top a := ⟨0, (mul_zero a).symm⟩
+
+instance : BoundedOrder (Associates α) where
+  -- { Associates.orderTop, Associates.orderBot with }
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:1013]:
```suggestion
```

(does that just work?)

File: Mathlib/Algebra/Associated.lean [hunk 11]
```diff
@@ -0,0 +1,1262 @@
...
+    simpa [is_unit_mk] using h (Associates.mk x) (Associates.mk y) rfl
+  · intro h x y
+    refine' Quotient.inductionOn₂ x y fun x y a_eq => _
+    rcases Quotient.exact a_eq.symm with ⟨u, a_eq⟩
+    rw [mul_assoc] at a_eq
+    show IsUnit (Associates.mk x) ∨ IsUnit (Associates.mk y)
+    simpa [is_unit_mk] using h _ _ a_eq.symm
+#align associates.irreducible_mk Associates.irreducible_mk
+
+theorem mk_dvd_not_unit_mk_iff {a b : α} :
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:1055]:
```suggestion
theorem mk_dvdNotUnit_mk_iff {a b : α} :
```

File: Mathlib/Algebra/Associated.lean [hunk 12]
```diff
@@ -0,0 +1,1262 @@
...
+#align associates.irreducible_iff_prime_iff Associates.irreducible_iff_prime_iff
+
+end CommMonoidWithZero
+
+section CancelCommMonoidWithZero
+
+variable [CancelCommMonoidWithZero α]
+
+instance : PartialOrder (Associates α) where
+  -- { Associates.preorder with
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:1101]:
```suggestion
```

File: Mathlib/Algebra/Associated.lean [hunk 13]
```diff
@@ -0,0 +1,1262 @@
...
+    le_antisymm := fun a' b' =>
+      Quotient.inductionOn₂ a' b' fun _ _ hab hba =>
+        Quot.sound <| associated_of_dvd_dvd (dvd_of_mk_le_mk hab) (dvd_of_mk_le_mk hba)
+
+instance : OrderedCommMonoid (Associates α) where
+  -- { Associates.commMonoid, Associates.partialOrder with
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:1107]:
```suggestion
```

File: Mathlib/Algebra/Associated.lean [hunk 14]
```diff
@@ -0,0 +1,1262 @@
...
+  | ⟨d, hd⟩ => ⟨d, eq_of_mul_eq_mul_left a _ _ ha <| by rwa [← mul_assoc]⟩
+#align associates.le_of_mul_le_mul_left Associates.le_of_mul_le_mul_left
+
+theorem one_or_eq_of_le_of_prime : ∀ p m : Associates α, Prime p → m ≤ p → m = 1 ∨ m = p
+  | p, m, ⟨hp0, _, h⟩, ⟨d, r⟩ => by
+    have dvd_rfl' : p ∣ m * d := by rw[r]
+    rw [r]
+    match h m d dvd_rfl' with
+    | Or.inl h' =>
+      -- (by_cases  m = 0 => by simp [this]) fun this : m ≠ 0 => by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:1140]:
```suggestion
```

File: Mathlib/Algebra/Associated.lean [hunk 15]
```diff
@@ -0,0 +1,1262 @@
...
+      by_cases m = 0
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:1141]:
```suggestion
      by_cases h : m = 0
```

File: Mathlib/Algebra/Associated.lean [hunk 16]
```diff
@@ -0,0 +1,1262 @@
...
+      case pos =>
+        simp [h, zero_mul]
+      case neg =>
+        rw [r] at h'
+        have : m * d ≤ m * 1 := by simpa using h'
+        have : d ≤ 1 := Associates.le_of_mul_le_mul_left m d 1 ‹m ≠ 0› this
+        have : d = 1 := bot_unique this
+        simp [this]
+    | Or.inr h' =>
+      -- (by_cases fun this : d = 0 => by simp [this] at hp0 <;> contradiction) fun this : d ≠ 0 =>
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:1151]:
```suggestion
```

File: Mathlib/Algebra/Associated.lean [hunk 17]
```diff
@@ -0,0 +1,1262 @@
...
+theorem le_one_iff {p : Associates α} : p ≤ 1 ↔ p = 1 := by rw [← Associates.bot_eq_one, le_bot_iff]
+#align associates.le_one_iff Associates.le_one_iff
+
+end CancelCommMonoidWithZero
+
+end Associates
+
+section CommMonoidWithZero
+
+theorem DvdNotUnit.is_unit_of_irreducible_right [CommMonoidWithZero α] {p q : α}
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Associated.lean:1186]:
```suggestion
theorem DvdNotUnit.isUnit_of_irreducible_right [CommMonoidWithZero α] {p q : α}
```

**Review by Ruben-VandeVelde — CHANGES_REQUESTED** [review]
Needs another careful review of the names.

File: Mathlib/Algebra/Associated.lean (hunk 10 above)
Comment (siddhartha-gadgil) [reply Mathlib/Algebra/Associated.lean:1011]: I get an error "expected command"

File: Mathlib/Algebra/Associated.lean (hunk 11 above)
Comment (siddhartha-gadgil) [reply Mathlib/Algebra/Associated.lean:1053]: Done Thanks

File: Mathlib/Algebra/Associated.lean (hunk 4 above)
Comment (siddhartha-gadgil) [reply Mathlib/Algebra/Associated.lean:283]: Done thansk

File: Mathlib/Algebra/Associated.lean (hunk 5 above)
Comment (siddhartha-gadgil) [reply Mathlib/Algebra/Associated.lean:382]: This was from mathport. Should it be deleted?

File: Mathlib/Algebra/Associated.lean (hunk 2 above)
Comment (siddhartha-gadgil) [reply Mathlib/Algebra/Associated.lean:173]: Done thanks

File: Mathlib/Algebra/Associated.lean (hunk 9 above)
Comment (siddhartha-gadgil) [reply Mathlib/Algebra/Associated.lean:996]: deleted

File: Mathlib/Algebra/Associated.lean [hunk 18]
```diff
@@ -0,0 +1,1262 @@
...
+#align associates.one_or_eq_of_le_of_prime Associates.one_or_eq_of_le_of_prime
+
+instance : CancelCommMonoidWithZero (Associates α) :=
+  { (inferInstance : CommMonoidWithZero (Associates α)) with
+    mul_left_cancel_of_ne_zero := by apply eq_of_mul_eq_mul_left
+    mul_right_cancel_of_ne_zero := by apply eq_of_mul_eq_mul_right }
+
+instance : CanonicallyOrderedMonoid (Associates α) where
+    exists_mul_of_le := by intro a b h; exact h
+    le_self_mul := fun a b => ⟨b, rfl⟩
```
[the remark this answers is not in the document]
Comment (siddhartha-gadgil) [reply Mathlib/Algebra/Associated.lean:1168]: I get an error:
```lean
type mismatch
  fun x x_1 h ↦ h
has type
  (x : a✝ ≤ b✝) → (x_1 : ?m.368054 x) → ?m.368055 x x_1 → ?m.368055 x x_1 : Sort (imax ?u.368048 ?u.368051)
but is expected to have type
  a✝ ≤ b✝ → Exists fun c ↦ b✝ = a✝ * c : Prop
the following variables have been introduced by the implicit lambda feature
  a✝ : Associates α
  b✝ : Associates α
```

File: Mathlib/Algebra/Associated.lean (hunk 16 above)
Comment (siddhartha-gadgil) [reply Mathlib/Algebra/Associated.lean:1149]: Done thanks

File: Mathlib/Algebra/Associated.lean [hunk 19]
```diff
@@ -0,0 +1,1262 @@
...
+        simp [h, zero_mul]
+      case neg =>
+        rw [r] at h'
+        have : m * d ≤ m * 1 := by simpa using h'
+        have : d ≤ 1 := Associates.le_of_mul_le_mul_left m d 1 ‹m ≠ 0› this
+        have : d = 1 := bot_unique this
+        simp [this]
+    | Or.inr h' =>
+      -- (by_cases fun this : d = 0 => by simp [this] at hp0 <;> contradiction) fun this : d ≠ 0 =>
+        by_cases d = 0
```
[the remark this answers is not in the document]
Comment (siddhartha-gadgil) [reply Mathlib/Algebra/Associated.lean:1150]: Done thanks

File: Mathlib/Algebra/Associated.lean [hunk 20]
```diff
@@ -0,0 +1,1262 @@
...
+  { (inferInstance : CommMonoidWithZero (Associates α)) with
+    mul_left_cancel_of_ne_zero := by apply eq_of_mul_eq_mul_left
+    mul_right_cancel_of_ne_zero := by apply eq_of_mul_eq_mul_right }
+
+instance : CanonicallyOrderedMonoid (Associates α) where
+    exists_mul_of_le := by intro a b h; exact h
+    le_self_mul := fun a b => ⟨b, rfl⟩
+    bot_le := by apply one_le
+
+theorem dvd_not_unit_iff_lt {a b : Associates α} : DvdNotUnit a b ↔ a < b :=
```
[the remark this answers is not in the document]
Comment (siddhartha-gadgil) [reply Mathlib/Algebra/Associated.lean:1171]: Done thanks

File: Mathlib/Algebra/Associated.lean [hunk 21]
```diff
@@ -0,0 +1,1253 @@
...
+    refine' ⟨i + 1, Nat.succ_le_succ hi, (hq.mul_left p).trans _⟩
+    rw [pow_succ]
+    rfl
+  · obtain ⟨i, hi, hq⟩ := ih.mp hno
+    exact ⟨i, hi.trans n.le_succ, hq⟩
+#align dvd_prime_pow dvd_prime_pow
+
+end CancelCommMonoidWithZero
+
+-- assert_not_exists multiset
```
Comment (semorrison) [inline Mathlib/Algebra/Associated.lean:1253]:
```suggestion
-- Porting note: `assert_not_exists` has not been ported yet.
-- assert_not_exists multiset
```

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**